src/utils.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright 2022 Aztec
+// Copyright 2025 Horizen Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0 or MIT
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright 2022 Aztec
+// Copyright 2024 Horizen Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0 or MIT
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    // errors::{FieldError, GroupError},
+    errors::FieldError,
+    types::G1,
+    Fq,
+    Fq2,
+    Fr,
+    G2,
+    U256,
+};
+use ark_bn254_ext::CurveHooks;
+use ark_ec::AffineRepr;
+use ark_ff::{AdditiveGroup, PrimeField};
+
+pub(crate) trait IntoFq {
+    fn into_fq(self) -> Fq;
+}
+
+impl IntoFq for U256 {
+    fn into_fq(self) -> Fq {
+        Fq::from(self)
+    }
+}
+
+// pub(crate) trait IntoFr {
+//     fn into_fr(self) -> Fr;
+// }
+
+// impl IntoFr for &[u8; 32] {
+//     fn into_fr(self) -> Fr {
+//         self.into_u256().into_fr()
+//     }
+// }
+
+// impl IntoFr for [u8; 32] {
+//     fn into_fr(self) -> Fr {
+//         (&self).into_fr()
+//     }
+// }
+
+// impl IntoFr for U256 {
+//     fn into_fr(self) -> Fr {
+//         Fr::new(self)
+//     }
+// }
+
+pub(crate) trait IntoU256 {
+    fn into_u256(self) -> U256;
+}
+
+impl IntoU256 for &[u8; 32] {
+    fn into_u256(self) -> U256 {
+        // Convert the byte array to a little-endian byte vector
+        let mut bytes = self.to_vec(); // Convert the &[u8; 32] slice to a Vec<u8>
+        bytes.reverse(); // Reverse the bytes to ensure little-endian order
+
+        // Create a BigInteger256 from the little-endian byte array
+        let mut limbs = [0u64; 4];
+
+        // Populate the limbs from the byte vector (which is little-endian)
+        for i in 0..4 {
+            limbs[i] = u64::from_le_bytes(
+                bytes[(i << 3)..((i + 1) << 3)]
+                    .try_into()
+                    .expect("Invalid byte slice"),
+            );
+        }
```

Comment:
Equivalent, more concise implementation

```suggestion
        let mut rchunks_iter = self.rchunks_exact(8);
        let limbs: [_; 4] = core::array::from_fn(|_| {
            u64::from_be_bytes(rchunks_iter.next().unwrap().try_into().unwrap())
        });
        debug_assert!(rchunks_iter.remainder().is_empty());
```